Backend/Backend/Services/ReportService.cs

```diff
@@ -0,0 +1,148 @@
+using Backend.Infrastructure;
+using Backend.Models;
+using Backend.Models.Interfaces;
+using Microsoft.EntityFrameworkCore;
+
+namespace Backend.Services;
+
+public class ReportService
+{
+    private readonly AppDbContext _db;
+    private readonly ILogger<ReportService> _logger;
+    private static readonly HashSet<string> AcceptableReportTargets = new() {
+        nameof(User),
+        nameof(Podcast),
+        nameof(Episode),    
+        nameof(Comment),
+        nameof(CommentReply)
+    };
+    
+    public ReportService(AppDbContext db, ILogger<ReportService> logger) {
+        _db = db;
+        _logger = logger;
+    }
+    
+    public async Task<ISoftDeletable> Report(Report report) {
+        // Check if entity is valid 
+        if (!AcceptableReportTargets.Contains(report.TargetEntityName)) {
+            throw new InvalidOperationException("Invalid report target");
+        }
+        
+        // Check that reported by is valid 
+        if (await _db.Users.FindAsync(report.ReportedBy) is null) {
+            throw new InvalidOperationException("Invalid reporter");
+        }
+        
+        // Check that target entity exists
+        var target = await GetReportTargetFromId(report.TargetEntityName, report.TargetId);
+        if (target is null) {
+            throw new InvalidOperationException("Invalid target entity");
+        }
+        
+        // Check if report already exists
+        if (await _db.Reports.AnyAsync(r => r.TargetEntityName == report.TargetEntityName && r.TargetId == report.TargetId && r.ReportedBy == report.ReportedBy)) {
+            throw new InvalidOperationException("Report already exists");
+        }
+        await _db.Reports.AddAsync(report);
+        _db.Reports.Add(report);
+        await _db.SaveChangesAsync();
+
+        return target;
+    }
+    
+    public Task<ImmutableArray<Report>> GetPendingReports() {
+        return Task.Run(() => _db.Reports.ToImmutableArray()) ;
+    }
+    
+    public async Task ResolveReport(User admin, Guid reportId) {
+        Report? report = await _db.Reports.FindAsync(reportId);
+        if (report is null) {
+            throw new InvalidOperationException("Invalid report ID");
+        }
+        await DeleteReportedEntity(admin.Id, report.TargetEntityName, report.TargetId);
+        report.Status = Models.Report.ReportStatus.Resolved;
+        report.DeletedBy = admin.Id;       
+        _db.Reports.Remove(report); // Soft deleted
+        await _db.SaveChangesAsync();
+    }
+    
+    public async Task RejectReport(User admin, Guid reportId) {
+        Report? report = await _db.Reports.FindAsync(reportId);
+        if (report is null) {
+            throw new InvalidOperationException("Invalid report ID");
+        }
+        report.Status = Models.Report.ReportStatus.Rejected;
+        report.DeletedBy = admin.Id;       
+        _db.Reports.Remove(report); // Soft deleted report
+        await _db.SaveChangesAsync();
+    }
+    
```

Comment:
You might want to seperate your Private and public methods with a \#region private methods, but not a huge deal